### ISSUE TITLE
Testsuite failure zig 0.10.0 in `behavior-native-Debug-c-multi-stage2_c`

### ISSUE THREAD

**Issue by freswa** [ziglang/zig#14240]
Hey there,
I'm about to bump the official zig package in Arch Linux to 0.10.0 since we'll finally introduce llvm 15 shortly.
Unfortunately while running the testsuite (in a clean environment) with
```
./zig2 build test -Dstatic-llvm=false -Denable-llvm=true -Dskip-non-native=true
```
I face the following error:
```
/build/zig/src/zig-0.10.0/zig-cache/o/7dd02ac3484d95bd7d039e5aa0dfa0f8/test.c:1:1: error: unable to build C object: clang exited with code 1
error: the following test command failed with exit code 1:
/build/zig/src/zig-0.10.0/build/zig2 run -lc /build/zig/src/zig-0.10.0/zig-cache/o/7dd02ac3484d95bd7d039e5aa0dfa0f8/test.c
error: test...
error: The following command exited with error code 1:
/build/zig/src/zig-0.10.0/build/zig2 test -fno-stage1 -fno-LLVM -ofmt=c /build/zig/src/zig-0.10.0/test/behavior.zig -lc --test-name-prefix behavior-native-Debug-c-multi-stage2_c  --cache-dir /build/zig/src/zig-0.10.0/zig-cache --global-cache-dir /build/.cache/zig --name test -fno-single-threaded -I /build/zig/src/zig-0.10.0/test --zig-lib-dir /build/zig/src/zig-0.10.0/lib --enable-cache
error: the following build command failed with exit code 1:
/build/zig/src/zig-0.10.0/zig-cache/o/e852fc47877d9b56c41d38688e78fbad/build /build/zig/src/zig-0.10.0/build/zig2 /build/zig/src/zig-0.10.0 /build/zig/src/zig-0.10.0/zig-cache /build/.cache/zig test -Dstatic-llvm=false -Denable-llvm=true -Dskip-non-native=true
```
I tried to debug this issue with the given source code, but failed to find any reason or solution (besides disabling the test).
I'd be glad if you would have a pointer for me how to tackle this.

Thanks

**Comment by andrewrk** [comment]
the stage2 binary is not intended to pass the test suite. you need to complete the build process, producing a stage3 binary, and then you can run the test suite.

**Comment by freswa** [comment]
This is my build step:
```
cmake -B build \
  -DCMAKE_BUILD_TYPE=None \
  -DZIG_SHARED_LLVM=ON \
  -DCMAKE_INSTALL_PREFIX=/usr \
  .
#make -C build
cmake --build build
```
Producing:
```
[...]
[100%] Built target zig2
[100%] Building stage3
[100%] Built target stage3
```

Though, while it promises stage3, there is no `stage3` directory at all. I replaced `cmake --build build` with `make -C build`, but I get the same behavior.

**Comment by andrewrk** [comment]
doesn't archlinux  have a standard way of building with cmake? you seem to have forgotten to run the install target.

**Comment by freswa** [comment]
Is `install` supposed to compile `stage3` then? This is the standard way and we usually run the tests before running the `install` step to aid reproducible builds.

**Comment by andrewrk** [comment]
No, `install` does not compile anything, it merely copies the build artifacts into the output prefix.

It sounds like what you want is for zig's cmake script to expose a `test` target, which can run the tests via the stage3 binary directly from the build artifact directory.

This is something I will look into for the next version, but it is not available in 0.10.x. I suggest to use the `install` target with a `--prefix` argument set to the build directory, and use this version for whatever testing you need, and then repeat your `install` again in your actual installation step, with the real prefix. The second install command will use cached artifacts and therefore not repeat any compilation.

**Comment by freswa** [comment]
I just found the stage 3 binary under `./zig-cache/o/dfa000634ac6023ce3d9f9c37eea4aca/zig`. It would be very helpful if this artifact would end up next to the `zig1` and `zig2` binaries in the build directory.

**Comment by BratishkaErik** [comment]
`zig build compile` in CMakeLists.txt can be changed to `zig build install`, then process should work like this:
 * `cmake build`
 * Stage 3 compiled
 * `zig build install` into default directory (IIRC directory-where-building/zig-out/
 * smth needed for distros (testing for example, or slotting)
 * (if `cmake install` used) `zig build install` for installing to actual directory

But it will require 2 copy operstions instead of one

**Comment by alexrp** [comment]
We are moving away from CMake once we drop the libLLVM dependency (https://github.com/ziglang/zig/issues/25492).
